source/http_stream.c

```diff
@@ -51,148 +50,163 @@ struct aws_http_stream *aws_py_get_http_stream(PyObject *stream) {
     return native;
 }
 
-static enum aws_http_outgoing_body_state s_stream_outgoing_body(
-    struct aws_http_stream *internal_stream,
-    struct aws_byte_buf *buf,
+static int s_on_incoming_headers(
+    struct aws_http_stream *native_stream,
+    const struct aws_http_header *header_array,
+    size_t num_headers,
     void *user_data) {
-    (void)internal_stream;
 
     struct http_stream_binding *stream = user_data;
+    int aws_result = AWS_OP_SUCCESS;
 
+    /*************** GIL ACQUIRE ***************/
     PyGILState_STATE state = PyGILState_Ensure();
 
-    PyObject *mv = aws_py_memory_view_from_byte_buffer(buf, PyBUF_WRITE);
+    /* Set this just in case callback fires before aws_http_connection_make_request() has even returned */
+    stream->native = native_stream;
 
-    if (!mv) {
-        aws_http_connection_close(aws_http_stream_get_connection(stream->native));
-        PyGILState_Release(state);
-        return AWS_HTTP_OUTGOING_BODY_DONE;
-    }
+    /* Build up list of (name,value) tuples as headers come in via repeated on_headers callacks. */
+    for (size_t i = 0; i < num_headers; ++i) {
+        const struct aws_http_header *header = &header_array[i];
+
+        PyObject *name_value_pair = Py_BuildValue(
+            "(s#s#)",
+            (const char *)header->name.ptr,
+            header->name.len,
+            (const char *)header->value.ptr,
+            header->value.len);
+        if (!name_value_pair) {
+            aws_result = aws_raise_py_error();
+            goto done;
+        }
 
-    PyObject *result = PyObject_CallFunction(stream->on_read_body, "(O)", mv);
-    /* the return from python land is a tuple where the first argument is the body state
-     * and the second is the amount written */
-    PyObject *state_code = PyTuple_GetItem(result, 0);
-    enum aws_http_outgoing_body_state body_state = (enum aws_http_outgoing_body_state)PyIntEnum_AsLong(state_code);
-    PyObject *written_obj = PyTuple_GetItem(result, 1);
-    long written = (long)PyLong_AsLong(written_obj);
-    Py_XDECREF(result);
-    Py_XDECREF(mv);
+        bool append_success = PyList_Append(stream->received_headers, name_value_pair) == 0;
```

Comment:
Interesting. Just out of curiosity, why not buffer in native and deliver all of the headers at once when the block is done, and not go over the VM barrier for each chunk?

Reply:
That is exactly what we're doing

Reply:
You make an excellent point. You could argue that's how the aws-c-http API should work, buffering response headers for you until they're all ready. I'll think about this...

Reply:
did it
we also discussed making an `aws_http_headers` datastructure that we could use in cases like this, outside of `aws_http_message`